contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,266 @@
+        // number == 0 means claim all
+        if (number == 0) {
+            number = _unbondRequestsQueue[_delegator].length();
+        }
+        if (number > _unbondRequestsQueue[_delegator].length()) {
+            number = _unbondRequestsQueue[_delegator].length();
+        }
+        require(number <= MAX_CLAIM_NUMBER, "TOO_MANY_REQUESTS"); // prevent too many loop in one transaction
+
+        uint256 _totalShares;
+        uint256 _totalBnbAmount;
+        while (number != 0) {
+            bytes32 hash = _unbondRequestsQueue[_delegator].front();
+            UnbondRequest memory request = _unbondRequests[hash];
+            if (block.timestamp < request.unlockTime) {
+                break;
+            }
+
+            _totalShares += request.sharesAmount;
+            _totalBnbAmount += request.bnbAmount;
+
+            // remove from the queue
+            _unbondRequestsQueue[_delegator].popFront();
+            delete _unbondRequests[hash];
+
+            number -= 1;
+        }
+
+        _lockedShares[_delegator] -= _totalShares;
+        uint256 gasLimit = IStakeHub(STAKE_HUB_ADDR).transferGasLimit();
+        (bool success,) = _delegator.call{value: _totalBnbAmount, gas: gasLimit}("");
+        require(success, "CLAIM_FAILED");
+
+        emit UnbondClaimed(_delegator, _totalShares, _totalBnbAmount);
+        return _totalBnbAmount;
+    }
+
+    function distributeReward() external payable onlyStakeHub {
+        uint256 _bnbAmount = msg.value;
+        _totalPooledBNB += _bnbAmount;
+        emit RewardReceived(_bnbAmount);
+    }
+
+    function slash(uint256 _slashBnbAmount) external onlyStakeHub returns (uint256) {
+        uint256 _securityDeposit = balanceOf(STAKE_HUB_ADDR);
+        uint256 _slashShares = getSharesByPooledBNB(_slashBnbAmount);
+
+        uint256 _remain;
+        if (_slashShares <= _securityDeposit) {
+            _totalPooledBNB -= _slashBnbAmount;
+            _burn(STAKE_HUB_ADDR, _slashShares);
+            _remain = 0;
+        } else {
+            uint256 _securityDepositBNB = getPooledBNBByShares(_securityDeposit);
+            _totalPooledBNB -= _securityDepositBNB;
+            _burn(STAKE_HUB_ADDR, _securityDeposit);
+
+            _remain = _slashBnbAmount - _securityDepositBNB;
+
+            _freeze = true;
+            _remainingSlashBnbAmount += _remain;
+        }
+
+        uint256 _realSlashBnbAmount = _slashBnbAmount - _remain;
+        (bool success,) = SYSTEM_REWARD_ADDR.call{value: _realSlashBnbAmount}("");
+        require(success, "TRANSFER_FAILED");
+        return _realSlashBnbAmount;
+    }
+
+    function lockToGovernance(address from, uint256 _sharesAmount) external onlyStakeHub returns (uint256) {
+        _transfer(from, GOVERNANCE_ADDR, _sharesAmount);
+        return getPooledBNBByShares(_sharesAmount);
+    }
+
+    function payFine() external payable {
+        require(_freeze, "NOT_FROZEN");
+        require(msg.sender == validator, "NOT_VALIDATOR");
```

Comment:
what about allow anyone to pay the fine?

Reply:
Done